datamade/just-spaces
# Permissioning system

## Overview

Closes #172. This PR implements three classes of users:

1. `Field users` can run published surveys.
2. `Staff users` have all the permissions of field users. They can also create, edit, and delete Agencies, Studies, Study Areas, Surveys, and Locations. They can publish surveys so they can be run by field users. Staff users can also view all collected data and design data visualizations on the collected data pages.
3. `Superusers` have all the permissions of staff users. They can also create, edit, and delete other users. Only superusers can set and change user passwords. **All superusers should also have staff status.**

@jeancochrane two notes:

1. I put in basic test coverage, but it could certainly be more extensive. Do you think that's worth doing here? 
2. Selecting both "Staff status" and "Superuser status" is confusing, and I'm not sure how best to explain how this works for users without creating new templates. What do you think?

## Testing Instructions

 * If you don't already have one, create a new superuser from the command line
* Once logged in as that superuser, go to `http://127.0.0.1:8000/admin/users/justspacesuser/`
* Create two new users: `fielduser`, who is neither a staff user or a superuser, and `staffuser`, who is both
* Log in as `fielduser`. Make sure that all you can do is select and run surveys
* Navigate to `http://127.0.0.1:8000/agencies` or another protected page. Make sure you're redirected to a login screen
* Log in as `staffuser`. Make sure you can see and do all of the above actions. 
* Create a new user as `staffuser`. Make sure that behaves as expected.
